design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+* lakectl command: a string with the `lakectl` command to test.
+* expected result: the expected output of a correct execution of the above command. This can be a string or, better, a file containing the output string
+
+In the future we might consider expending this structure with more fields, to allow extended functionality
+ 
+### How should we use `lakefs` backend
+
+Here are some of the pros and cons of each approach
+
+#### **Live Server**
+
+On the upside:
+* Straightforward - the implementation is already there, just bring the server up.
+  * Moreover, `lakefs` modifications are immediately reflected into the tests
+* The best possible simulation - this is the real server `lakectl` is communicating with
+
+And then, on the downside:
+* Bringing up the server requires additional dependencies - underlying storage, DB etc. While this supports the advantage of testing a **real** server environment, it complicates the test setup, and although this is easily solved by containers, it still worth mentioning
+* Performance-wise, creating a server environment for each test (thousands of them, ideally) is very expensive. We can, of course, group several tests (all of them?) to use a single server session, but this will eliminate the isolative and stateless nature of the tests, which is undesired
+* Flexibility is limited to `lakefs` server implementation. This means that testing `lakectl` behavior, with improbable response from `lakefs` server (e.g. timeout, disconnection etc.) becomes very difficult
+
+#### Mock
+
+Advantages:
+* Totally independent from the environment. No need for DB, underlying storage, or any other future dependency that may be added to `lakefs` server. Moreover, future removal of any dependency from `lakefs` (Postgres?!) will not require framework modification. And, once again, although this advantage is pretty much diminished by the usage of containers, it is still worth noting
+* No uptime, as the mock is a part of the test itself
+* Totally stateless - again, as the mock is part of each test
+* Maximal flexibility - each desired server response can be easily hard-coded
+* No need for initial state setup steps
+* Server bugs are not included
+
+Disadvantages:
+* Requires implementation - unlike `lakefs` server, we currently don't have a mock
+* Requires (a lot of) maintenance - every new feature and every bug fix in `lakefs` may, and probably will, need to be reflected in the mock
+
+#### Conclusion ?
+
+For the long term it look like the better approach is to use a full scale mock, which should evolve as tests are created. While this will certainly have an impact on new tests creation ease, the benefits of lack independency on `lakefs` are too important to ignore, especially when dealing with a young product such as `lakefs` which may be subdued to significant changes in the future
```

Comment:
_Commented by @itaiad200_
After reading the 2 options, I ended up with a different conclusion smile
IMO a full scale mock will end up with writing a parallel mocked lakeFS implementation, which will be hard to maintain, expand and debug. Running a real lakeFS instance means that we would be testing it too, which from the scope of lakectl testing sounds like a downside, but lakectl always interacts with a lakeFS server, so we really need to test the complete lakectl -> lakeFS flow.

Reply:
_Thumbed up by @johnnyaug_
:+1:

Reply:
If so, let's change the scope of this ticket to end-2-end tests which will cover client to server to client. We can design it to verify results at each step of the way, i.e. verify the requests sent to the server. verify the response returned from the server and verify the client output (no-output expected is also verifiable)
We can support multi-action or multi-command flows, making the dependencies a desirable side effect, with a single command test being a trivial case of the above.
We can start with `lakectl` as the client to use, but design the framework to be extensible to other clients (WebUI, WebAPI etc)
**And** for quick development of `lakectl` tests, using mocked environment, we can start pushing [lakectl usin tests](https://github.com/treeverse/lakeFS/issues/2499)

Reply:
I don't think that we should change the scope of it from `lakectl` testing to `end-2-end` testing since we already have that. Reading thru my comments I understand why it sounds like I'm pushing for a `end-2-end` testing :)  see my previous comment ..